Document supertype(T::UnionAll) (#54286)

xlxs4 opened the pull request:
Fixes https://github.com/JuliaLang/julia/issues/54286.

---

Does this seem reasonable, or is it too minimal, considering that people using `supertype` might not have any idea what `UnionAll`s are, and it should at least point to their (developer) documentation, or explain them further?

xlxs4 committed:
Document supertype(T::UnionAll)

Keno committed:
interpreter: Use world-age-partitioned cache for @generated results (#54362)

This fixes #54360 by moving the interpreter's cache of `@generated`
results from `mi.uninferred` into `mi.cache` with a separate cache owner
to partition the cache from regular inference results. There are two
other uses of the `mi.uninferred` field:

1. As the place to store uninferred code for temporary top-level thunks
2. Is an uncompressed copy of m->source to avoid having to re-uncompress
every time in the interpreter.

In this PR, use case 1 is changed to use the same mechanism as generated
functions. Use case 2 is changed to just uncompress the source in place
in m->source. As a result, the `uninferred` field is unused and removed.

Note that I'm planning a somewhat larger refactor of `MethodInstance` in
the immediate future, so this might be a somewhat shortlived
representation, but that change should hopefully by largely transparent
to users of the wrappers introduced here.

xlxs4 committed:
Document copy! for vector types (#54142) (#54392)

Fixes https://github.com/JuliaLang/julia/issues/54142. Also documents
the `firstindex(dst) == firstindex(src)` check, since the `axes(dst) ==
axes(src)` is already documented.

topolarity committed:
specialize `VersionNumber` constructors less aggressively (#54386)

There's no need to specialize this code for every set of Tuple types
that it might see, especially when the primary callers don't know their
Tuple type any more precisely than this anyway.

Eliminates a dynamic dispatch from `tryparse(VersionNumber, "...")`

Co-authored-by: Gabriel Baraldi <baraldigabriel@gmail.com>

jishnub committed:
Unicode: assume foldable ccall in category_code (#54394)

Following on from https://github.com/JuliaLang/julia/pull/54346, this
marks the `ccall` in `category_code` as foldable. This lets us compute
the results of several functions at compile time, such as:
```julia
julia> @code_typed (() -> isletter('C'))()
CodeInfo(
1 ─     return true
) => Bool

julia> @code_typed (() -> isnumeric('C'))()
CodeInfo(
1 ─     return false
) => Bool

julia> @code_typed (() -> ispunct('C'))()
CodeInfo(
1 ─     return false
) => Bool
```

dkarrasch committed:
Stabilize `MulAddMul` strategically (#52439)

Co-authored-by: Ashley Milsted <ashmilsted@gmail.com>

KristofferC committed:
fix `code_warntype` on `OpaqueClosure`s (#54378)

Fixes https://github.com/JuliaLang/julia/issues/54375

---------

Co-authored-by: Shuhei Kadowaki <aviatesk@gmail.com>

nsajko commented on base/operators.jl:
```diff
@@ -68,12 +68,16 @@ Supertype operator, equivalent to `T2 <: T1`.
 """
     supertype(T::DataType)
```
Keep this consistent with the text below. So it should be, e.g., `supertype(T::Union{DataType,UnionAll})`, I guess.

xlxs4 replied:
I think usually when there's more than one method there's no type annotation? No idea ^^

nsajko replied:
IMO it would maybe be best to put `supertype(T::Type)` here, but clarify in the description below that only UnionAll and DataType are accepted. You could also explicitly tell that `Union`s aren't allowed.

nsajko commented on base/operators.jl:
```diff
@@ -68,12 +68,16 @@ Supertype operator, equivalent to `T2 <: T1`.
     supertype(T::DataType)
 
-Return the supertype of DataType `T`.
+Return the supertype of type `T`.
```
Don't use the definite article here, a type may have more than one supertype. Not sure what the best wording is.

xlxs4 replied:
I've seen direct supertype around (e.g. [oracle docs](https://docs.oracle.com/javase/specs/jls/se21/html/jls-4.html)), would that work?

nsajko commented on base/operators.jl:
```diff
@@ -68,12 +68,16 @@ Supertype operator, equivalent to `T2 <: T1`.
 
-Return the supertype of DataType `T`.
+Return the supertype of type `T`.
+`T` can be a `DataType` or a `UnionAll` type.
```
Actually, this sentence isn't very useful, given that the allowed argument type should be indicated in the signature above anyway.

nsajko replied:
Although it could be good to xref `DataType` and `UnionAll`. Like this:

```md
[`DataType`](@ref)
```

```md
[`UnionAll`](@ref)
```

DilumAluthgeBot committed:
🤖 [master] Bump the SparseArrays stdlib from cb602d7 to a09f90b (#54406)

Stdlib: SparseArrays
URL: https://github.com/JuliaSparse/SparseArrays.jl.git
Stdlib branch: main
Julia branch: master
Old commit: cb602d7
New commit: a09f90b
Julia version: 1.12.0-DEV
SparseArrays version: 1.12.0
Bump invoked by: @dkarrasch
Powered by:
[BumpStdlibs.jl](https://github.com/JuliaLang/BumpStdlibs.jl)

Diff:
https://github.com/JuliaSparse/SparseArrays.jl/compare/cb602d7b7cf46057ddc87d23cda2bdd168a548ac...a09f90b43105229e42ac687ab4e3f6cf50deed8c

```
$ git log --oneline cb602d7..a09f90b
a09f90b Adjust matvec and matmatmul! to new internal LinAlg interface (#519)
3b30333 ci: run aqua test as a standalone ci job (#537)
df0a154 Add versioned Manifest files to .gitignore (#534)
4606755 Extend `copytrito!` for a sparse source (#533)
33fbc75 SparseMatrixCSC constructor with a Tuple of Integers (#523)
08d6ae1 CI: don't run `threads` tests in Windows GHA CI (attempt 2) (#530)
7408e4b Revert "Don't fail CI if codecov upload fails." (#527)
287e406 Bump julia-actions/setup-julia from 1 to 2 (#524)
b5de0da Don't fail CI if codecov upload fails. (#525)
78dde4c cast to Float64 directly instead of using float (#521)
a5e95ec CI: Add Apple Silicon (macOS aarch64) to the CI matrix (#505)
```

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

NeroBlackstone committed:
Add musical symbols for unicode completions (#54182)

I found that julia was missing some useful music notation unicode
completions and sometimes [had to copy and paste these unicodes to use
the published
package](https://github.com/JuliaMusic/MusicTheory.jl/issues/27).

So I picked out some unicode music symbols that might be useful, named
from the [unicode
standard](https://www.unicode.org/charts/PDF/U1D100.pdf).

I didn't add the test, I don't think it's necessary.

Thanks for reviewing. Please reply if any changes are needed.

---------

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

xlxs4 committed:
Further clarify supertype docs

xlxs4 commented:
I didn't force... I can re-open if needed

LilithHafner commented:
Probably worth re-opening. or force pushing back to b64c581

nsajko commented:
Simply run `git rebase -i master` on the `supertype-docs` branch, and drop the unnecessary commits by deleting the associated lines. Then force push after everything is fixed.
